### PR TITLE
Fix GC instability when using -d:useNimRtl (for the default gc)

### PR DESCRIPTION
**Pull request by idlewan**
## GC Problem when using `-d:useNimRtl`

When using `-d:useNimRtl`, the GC is quite unstable. A short example (compile it with GC asserts on) to demonstrate:
```nim
var s: seq[int] = newSeq[int](3)
s[0] = 1
s[1] = 2
s[2] = 3
echo s    # works

var s2: seq[int] = @[4, 5, 6]
echo "echoing s2"  # isn't printed, previous line fails in a GCASSERT when -d:useNimRtl
echo s2
```
Compile without nimrtl: 
`nim c -d:useGCAssert -d:useSysAssert -d:fulldebug -o:test_gc_normal test_gc.nim`

Compile with nimrtl:
`nim c -d:useGCAssert -d:useSysAssert -d:fulldebug -o:libnimrtl.so nim-repo/lib/nimrtl.nim`
`nim c -d:useGCAssert -d:useSysAssert -d:fulldebug -d:useNimRtl -o:test_gc_nimrtl test_gc.nim`

Run (on linux): 
`./test_gc_normal` (works)
```output
@[1, 2, 3]
echoing s2
@[4, 5, 6]
```
`LD_LIBRARY_PATH=/path/to/libnimrtl.so ./test_gc_nimrtl` (fails in a GC assert)
```output
@[1, 2, 3]
[GCASSERT] incRef: interiorPtr
No stack traceback available
```

This is when enabling asserts, but I've also experienced crashes when (as always, using nimrtl) loading a library with loadLib, allocating and printing strings (using `$` of a matrix in glm), unloading the library, loading the library again, and printing again (segfaulted in `gc.nim` in `growObj` or `newObj`).

## What happens when using `-d:useNimRtl`
The generated c files for `libnimrtl.so` contain, as expected, the variable `gch` of type `gcHeap`, and functions of `gc.nim`.
The generated c files for `test_gc_nimrtl` also contain, however, some of the same functions, and a `gch` variable. This is the source of the problem, it's a similar problem as #6716.

## Solution
Any function that reference the `gch` module-global variable in `gc.nim` and `gc_common.nim` should not have their implementation included in programs that use `-d:useNimRtl`. They should only exist in `libnimrtl.so`.
By extension, that also means every function that use `forAllChildren`, `isOnStack`, `doOperation`, and `collectZCT`, which all reference `gch` (except `collectZCT`, which reference `forAllChildren`).

### Current solution
This commit thus pushes all of those functions to be linked from the `libnimrtl.so` instead of being compiled in, with the help of the rtl helper pragmas. ~~It fixes the crashes I've for now experienced~~ it fixes the test program at the top of this post, but maybe I didn't get everything because I still seem to experience crashes for my program that load libs. I thought I solved it because I used `-d:fulldebug`, however I still experience crashes without.

### Better solution
For a safer, more permanent solution, this is what should probably be done in `gc.nim`:
```patch
-var
-  gch {.rtlThreadVar.}: GcHeap

when not defined(useNimRtl):
+ var
+    gch {.rtlThreadVar.}: GcHeap
  instantiateForRegion(gch.region)
```
and put under other `when not defined(useNimRtl):` switches the implementation of functions  `forAllChildren`, `isOnStack`, `doOperation`, and `collectZCT`, which could themselves then receive the `{.stl.}` pragma in their forward declarations.
Since this better solution would be a bigger diff, I prefered implementing the first solution (as it is shorter), describe the issue here and wait for a go to do the better solution.

### CODE REVIEW

**Comment by idlewan** [comment]
To note: this is a WIP, it probably fixes some things, but not everything yet.

File: lib/system/gc.nim [hunk 1]
```diff
@@ -263,7 +263,7 @@ proc nimGCunrefNoCycle(p: pointer) {.compilerProc, inline.} =
 proc nimGCunrefRC1(p: pointer) {.compilerProc, inline.} =
   decRef(usrToCell(p))
 
-proc asgnRef(dest: PPointer, src: pointer) {.compilerProc, inline.} =
+proc asgnRef(dest: PPointer, src: pointer) {.compilerRtl, inl.} =
```
Comment (Araq) [inline lib/system/gc.nim]: This is quite expensive. The code was written in order to support ``inline`` in DLL builds. Only procs that access ``gch`` directly need to go through the DLL indirection.

**Comment by idlewan** [comment]
I think my approach is wrong. I thought it would be the same type of problem as #6716, where things that should get only compiled in the  `nimrtllib.so` are accidentally duplicated in a program using it, leading to some functions operating on a duplicate variable. I thought this would be an easy fix, but I didn't succeed.

I'm using a variant of https://github.com/Serenitor/hotnim to test changes.
Somehow even when marking most if not all procs in `gc.nim` and `gc_common.nim`, I still get either a failed sysassert or a segfault.
To force a segfault to happen, I load a library with the `dynlib` module, echo/allocate some strings, do `GC_fullcollect`, unload the library, do another `GC_fullcollect`, and then the program segfaults. This happens with the default gc.
The workaround that I found is to only do a `GC_fullcollect` when the library is loaded, and to use `--gc:v2`. That seems to not segfault or fail on asserts, so there must be something happening in the default gc that leads to the gc trying to access something (probably a finalizer/destructor?) that belongs in the loaded library's address space. That's just my guess, I think I am out of my depth here.

I'm closing this, since I'm afraid I didn't find how to fix it, hopefully this gives some detail for the next person that wants to make the default gc work with `libnimrtl.so`.

**Comment by Araq** [comment]
Well if you ``unload``  the library all sort of problems might occur. Like the GC still trying to follow a "registered" root that doesn't exist anymore. I think that ``--gc:v2`` works was just luck and your analysis is close to nailing the real issue.

**Pull request closed without merging**